src/isolate.rs

```diff
@@ -344,38 +349,107 @@ impl Isolate {
+    Self(isolate)
   }
-}
 
-/// Internal method for constructing an OwnedIsolate.
-pub unsafe fn new_owned_isolate(isolate_ptr: *mut Isolate) -> OwnedIsolate {
-  OwnedIsolate(NonNull::new(isolate_ptr).unwrap())
-}
+  pub(crate) unsafe fn from_raw(self_ptr: *mut Isolate) -> Self {
+    Self(NonNull::new(self_ptr).unwrap())
+  }
 
-/// Same as Isolate but gets disposed when it goes out of scope.
-pub struct OwnedIsolate(NonNull<Isolate>);
+  pub(crate) fn into_raw(self) -> *mut Isolate {
+    let self_ptr = self.0.as_ptr();
+    forget(self);
+    self_ptr
+  }
 
-impl Drop for OwnedIsolate {
-  fn drop(&mut self) {
-    unsafe { self.0.as_mut().dispose() }
+  pub(crate) fn as_ptr(&self) -> *mut Isolate {
+    self.0.as_ptr()
+  }
+
+  /// Forcefully terminate the current thread of JavaScript execution
+  /// in the given isolate.
+  ///
+  /// This method can be used by any thread even if that thread has not
+  /// acquired the V8 lock with a Locker object.
+  pub fn terminate_execution(&self) {
+    unsafe { v8__Isolate__TerminateExecution(self.as_ptr()) }
+  }
+
+  /// Resume execution capability in the given isolate, whose execution
+  /// was previously forcefully terminated using TerminateExecution().
+  ///
+  /// When execution is forcefully terminated using TerminateExecution(),
+  /// the isolate can not resume execution until all JavaScript frames
+  /// have propagated the uncatchable exception which is generated.  This
+  /// method allows the program embedding the engine to handle the
+  /// termination event and resume execution capability, even if
+  /// JavaScript frames remain on the stack.
+  ///
+  /// This method can be used by any thread even if that thread has not
+  /// acquired the V8 lock with a Locker object.
+  pub fn cancel_terminate_execution(&self) {
+    unsafe { v8__Isolate__CancelTerminateExecution(self.as_ptr()) }
+  }
+
+  /// Request V8 to interrupt long running JavaScript code and invoke
+  /// the given |callback| passing the given |data| to it. After |callback|
+  /// returns control will be returned to the JavaScript code.
+  /// There may be a number of interrupt requests in flight.
+  /// Can be called from another thread without acquiring a |Locker|.
+  /// Registered |callback| must not reenter interrupted Isolate.
+  // Clippy warns that this method is dereferencing a raw pointer, but it is
+  // not: https://github.com/rust-lang/rust-clippy/issues/3045
+  #[allow(clippy::not_unsafe_ptr_arg_deref)]
+  pub fn request_interrupt(
+    &self,
+    callback: InterruptCallback,
+    data: *mut c_void,
+  ) {
+    unsafe { v8__Isolate__RequestInterrupt(self.as_ptr(), callback, data) }
   }
 }
 
-impl Deref for OwnedIsolate {
-  type Target = Isolate;
-  fn deref(&self) -> &Self::Target {
-    unsafe { self.0.as_ref() }
+impl Clone for IsolateHandle {
+  fn clone(&self) -> Self {
+    let prev_ref_count = unsafe { IsolateAnnex::get(&self.0) }
+      .get_ref_count()
+      .fetch_add(1, Ordering::Release);
```

Comment:
This can be `Ordering::Relaxed` and the fetch_sub() on line 446 `Ordering::AcqRel` (IOW, this one's too strong and the other one too weak.)

For comparison, [that's how std::shared_ptr is implemented in libc++](https://bugs.llvm.org/show_bug.cgi?id=22803).

Reply:
Good catch!